dbms/src/Storages/StorageMySQL.cpp

```diff
@@ -0,0 +1,187 @@
+        bool first = true;
+        for (const auto & expr: and_expression->arguments->children)
+        {
+            if (isCompatible(*expr.get()))
+            {
+                if (!first)
+                    stream << " AND ";
+                first = false;
+                dumpWhere(*expr.get(), stream);
+            }
+        }
+    }
+    else
+    {
+        if (isCompatible(where))
+        {
+            dumpWhere(where, stream);
+        }
+    }
+}
+
+/** Function analyze query builds select query of all used columns in query_info from table set by table_name parameter with where expression from filtered query_info.
+  * Also it is builds sample_block with all columns, where types are found in column_map
+  */
+
+std::string analyzeQuery(const SelectQueryInfo & query_info, const Context & context, std::string table_name, NamesAndTypesListPtr columns, google::dense_hash_map<std::string, DataTypePtr> & column_map, Block & sample_block)
+{
+    BlockInputStreams res;
+    StoragePtr storage(NULL);
+    ExpressionAnalyzer analyzer(query_info.query, context, storage, *columns);
+    NamesAndTypesList* usedColumns = analyzer.getColumns();
+    std::stringstream iss;
+    iss << "SELECT ";
+    bool first = true;
+    for (const auto & column: *usedColumns)
+    {
+        if (!first)
+            iss << ",";
+        iss << column.name;
+        first = false;
+        ColumnWithTypeAndName col;
+        col.name = column.name;
+        google::dense_hash_map<std::string, DataTypePtr>::iterator it = column_map.find(column.name);
+        if (it == column_map.end())
+        {
+            throw Exception("Can not find field " + column.name + " in table " + table_name, ErrorCodes::CANNOT_FIND_FIELD);
+        }
+        col.type = it->second;
+        col.column = column.type->createColumn();
+        sample_block.insert(std::move(col));
+    }
+    iss << " FROM " << table_name;
+    const ASTSelectQuery * select_query = typeid_cast<ASTSelectQuery *>(query_info.query.get());
+    const IAST * where = select_query->where_expression.get();
+    if (where)
+    {
+        std::stringstream where_stream;
+//        fprintf(stderr, "WHERE treeId: %s\n", where->getTreeID().c_str());
+        filterWhere(*where, where_stream);
+        std::string filtered_where = where_stream.str();
+        if (filtered_where.size())
+        {
+            iss << " WHERE " << filtered_where;
+        }
+//        fprintf(stderr, "Filtered WHERE: %s\n", filtered_where.c_str());
+    }
+    return iss.str();
+}
+
+BlockInputStreams StorageMySQL::read(
+    const Names & column_names,
+    const SelectQueryInfo & query_info,
+    const Context & context,
+    QueryProcessingStage::Enum & processed_stage,
+    size_t max_block_size,
+    unsigned num_streams)
+{
+    BlockInputStreams res;
+    sample_block.clear();
+    std::string query = analyzeQuery(query_info, context, mysql_table_name, columns, column_map, sample_block);
```

Comment:
`mysql_table_name, columns, column_map, sample_block` are class members. May be it's easy to make `analyzeQuery` private member-function?

Reply:
It Is also called from StorageODBC

Reply:
I've seen that `analyzeQuery` is used in `StorageODBC` also. May be move function declaration and implementation in separate files?